### PR TITLE
Update the MissionExecutor for lifecycle nodes

### PR DESCRIPTION
**Pull request by mjeronimo**
## Description
* Update the MissionExecutor for lifecycle nodes
* Add a test action client that can be used to send a mission plan to the mission executor.

## Future work
* Convert the test to a gtest-based unit test

### CODE REVIEW

File: nav2_mission_executor/src/mission_executor.cpp [hunk 1]
```diff
@@ -14,41 +14,123 @@
...
+
+  // Run the Behavior Tree
+  auto is_canceling = [goal_handle]() -> bool {return goal_handle->is_canceling();};
+  nav2_tasks::BtStatus rc = bt.run(blackboard, xml_string, is_canceling);
+
+  // Handle the result
+  switch (rc) {
+    case nav2_tasks::BtStatus::SUCCEEDED:
+      RCLCPP_INFO(get_logger(), "Mission succeeded");
+      goal_handle->set_succeeded(result);
```
Comment (bpwilcox) [inline nav2_mission_executor/src/mission_executor.cpp]: Same request I made in #651 about the renamed action state transitions, please change

File: nav2_mission_executor/src/mission_executor.cpp [hunk 2]
```diff
@@ -14,41 +14,123 @@
...
+  RCLCPP_INFO(get_logger(), "Creating");
+}
+
+MissionExecutor::~MissionExecutor()
+{
+  RCLCPP_INFO(get_logger(), "Destroying");
+}
+
+nav2_lifecycle::CallbackReturn
+MissionExecutor::on_configure(const rclcpp_lifecycle::State &)
```
Comment (mhpanah) [inline nav2_mission_executor/src/mission_executor.cpp]: Same as my previous comments, we've been using `\*...*\` for unused parameters.

File: nav2_mission_executor/src/mission_executor.cpp [hunk 3]
```diff
@@ -14,41 +14,123 @@
...
+
+  // Create the action server that we implement with our executeMission method
+  action_server_ = std::make_unique<ActionServer>(rclcpp_node_, "ExecuteMission",
+      std::bind(&MissionExecutor::executeMission, this, std::placeholders::_1));
+
+  return nav2_lifecycle::CallbackReturn::SUCCESS;
+}
+
+nav2_lifecycle::CallbackReturn
+MissionExecutor::on_activate(const rclcpp_lifecycle::State &)
```
Comment (mhpanah) [inline nav2_mission_executor/src/mission_executor.cpp]: Same as above.

File: nav2_mission_executor/src/mission_executor.cpp [hunk 4]
```diff
@@ -14,41 +14,123 @@
...
+{
+  RCLCPP_INFO(get_logger(), "Activating");
+  return nav2_lifecycle::CallbackReturn::SUCCESS;
+}
+
+nav2_lifecycle::CallbackReturn
+MissionExecutor::on_deactivate(const rclcpp_lifecycle::State &)
```
Comment (mhpanah) [inline nav2_mission_executor/src/mission_executor.cpp]: Same as above.

File: nav2_mission_executor/src/mission_executor.cpp [hunk 5]
```diff
@@ -14,41 +14,123 @@
...
+{
+  RCLCPP_INFO(get_logger(), "Deactivating");
+  return nav2_lifecycle::CallbackReturn::SUCCESS;
+}
+
+nav2_lifecycle::CallbackReturn
+MissionExecutor::on_cleanup(const rclcpp_lifecycle::State &)
```
Comment (mhpanah) [inline nav2_mission_executor/src/mission_executor.cpp]: same as above.

File: nav2_mission_executor/src/mission_executor.cpp [hunk 6]
```diff
@@ -14,41 +14,123 @@
...
 
-  task_server_ = std::make_unique<nav2_tasks::ExecuteMissionTaskServer>(temp_node);
+  client_node_.reset();
+  action_server_.reset();
+
+  return nav2_lifecycle::CallbackReturn::SUCCESS;
+}
+
+nav2_lifecycle::CallbackReturn
+MissionExecutor::on_error(const rclcpp_lifecycle::State &)
```
Comment (mhpanah) [inline nav2_mission_executor/src/mission_executor.cpp]: same.

File: nav2_mission_executor/src/mission_executor.cpp [hunk 7]
```diff
@@ -14,41 +14,123 @@
...
+{
+  RCLCPP_ERROR(get_logger(), "Handling error state");
+  return nav2_lifecycle::CallbackReturn::SUCCESS;
+}
 
-  task_server_->setExecuteCallback(
-    std::bind(&MissionExecutor::executeMission, this, std::placeholders::_1));
+nav2_lifecycle::CallbackReturn
+MissionExecutor::on_shutdown(const rclcpp_lifecycle::State &)
```
Comment (mhpanah) [inline nav2_mission_executor/src/mission_executor.cpp]: same.

File: nav2_mission_executor/test/test_mission_executor.cpp [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+  action_client->wait_for_action_server();
+
+  // The goal contains the XML representation of the BT
+  auto goal = nav2_msgs::action::ExecuteMission::Goal();
+  goal.mission_plan.mission_plan = xml_text;
+
+  // Send the goal
+  auto future_goal_handle = action_client->async_send_goal(goal);
+  if (rclcpp::spin_until_future_complete(node, future_goal_handle) !=
```
Comment (mhpanah) [inline nav2_mission_executor/test/test_mission_executor.cpp]: We should have a timeout here.

File: nav2_mission_executor/test/test_mission_executor.cpp [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+
+  auto goal_handle = future_goal_handle.get();
+  if (!goal_handle) {
+    RCLCPP_ERROR(node->get_logger(), "Goal was rejected by server");
+    return 1;
+  }
+
+  // Wait for the result
+  auto future_result = goal_handle->async_result();
+  if (rclcpp::spin_until_future_complete(node, future_result) !=
```
Comment (mhpanah) [inline nav2_mission_executor/test/test_mission_executor.cpp]: Timeout here as well.

File: nav2_mission_executor/src/mission_executor.cpp (hunk 2 above)
Comment (mjeronimo) [reply nav2_mission_executor/src/mission_executor.cpp]: Fixed

File: nav2_mission_executor/src/mission_executor.cpp (hunk 3 above)
Comment (mjeronimo) [reply nav2_mission_executor/src/mission_executor.cpp]: Fixed.

File: nav2_mission_executor/src/mission_executor.cpp (hunk 5 above)
Comment (mjeronimo) [reply nav2_mission_executor/src/mission_executor.cpp]: Fixed.

File: nav2_mission_executor/src/mission_executor.cpp (hunk 6 above)
Comment (mjeronimo) [reply nav2_mission_executor/src/mission_executor.cpp]: Fixed.

File: nav2_mission_executor/src/mission_executor.cpp (hunk 7 above)
Comment (mjeronimo) [reply nav2_mission_executor/src/mission_executor.cpp]: Fixed.

File: nav2_mission_executor/src/mission_executor.cpp (hunk 1 above)
Comment (mjeronimo) [reply nav2_mission_executor/src/mission_executor.cpp]: Updated.

File: nav2_mission_executor/test/test_mission_executor.cpp (hunk 8 above)
Comment (mjeronimo) [reply nav2_mission_executor/test/test_mission_executor.cpp]: I did this on purpose as I wanted to know if and exactly where there might be a hang. That is, I wanted to focus on reliability first before adding timeouts. We should discuss as a group.

File: nav2_mission_executor/test/test_mission_executor.cpp (hunk 9 above)
Comment (mjeronimo) [reply nav2_mission_executor/test/test_mission_executor.cpp]: See previous comment.

File: nav2_mission_executor/src/mission_executor.cpp (hunk 4 above)
Comment (mjeronimo) [reply nav2_mission_executor/src/mission_executor.cpp]: Fixed.

File: nav2_mission_executor/test/test_mission_executor.cpp (hunk 8 above)
Comment (mhpanah) [reply nav2_mission_executor/test/test_mission_executor.cpp]: Sure.

**Pull request merged**